azure/services/aso/service.go

```diff
@@ -0,0 +1,123 @@
+type Service[T deepCopier[T], S Scope] struct {
+	Reconciler[T]
+
+	Scope S
+	Specs []azure.ASOResourceSpecGetter[T]
+
+	PostCreateOrUpdateResourceHook func(scope S, result T, err error)
+	PostReconcileHook              func(scope S, err error) error
+	PostDeleteHook                 func(scope S, err error) error
+
+	name string
+}
+
+// NewService creates a new Service.
+func NewService[T deepCopier[T], S Scope](name string, scope S) *Service[T, S] {
+	return &Service[T, S]{
+		name:       name,
+		Scope:      scope,
+		Reconciler: New[T](scope.GetClient(), scope.ClusterName()),
+	}
+}
+
+// Name returns the service name.
+func (s *Service[T, S]) Name() string {
+	return s.name
+}
+
+// Reconcile idempotently creates or updates the resources.
+func (s *Service[T, S]) Reconcile(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "aso.Service.Reconcile")
+	defer done()
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	// We go through the list of Specs to reconcile each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one.
+	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
+	var resultErr error
+	for _, spec := range s.Specs {
+		result, err := s.CreateOrUpdateResource(ctx, spec, s.Name())
+		if err != nil && (!azure.IsOperationNotDoneError(err) || resultErr == nil) {
+			resultErr = err
+		}
+		if s.PostCreateOrUpdateResourceHook != nil {
+			s.PostCreateOrUpdateResourceHook(s.Scope, result, err)
+		}
+	}
+
+	if s.PostReconcileHook != nil {
+		return s.PostReconcileHook(s.Scope, resultErr)
+	}
+	return resultErr
+}
+
+// Delete deletes the resources.
+func (s *Service[T, S]) Delete(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "aso.Service.Delete")
+	defer done()
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	// We go through the list of SubnetSpecs to delete each one, independently of the resultErr of the previous one.
+	// If multiple errors occur, we return the most pressing one.
+	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error deleting) -> operationNotDoneError (i.e. deleting in progress) -> no error (i.e. deleted)
+	var resultErr error
+	for _, spec := range s.Specs {
+		err := s.DeleteResource(ctx, spec, s.Name())
+		if err != nil && (!azure.IsOperationNotDoneError(err) || resultErr == nil) {
+			resultErr = err
+		}
+	}
+
+	if s.PostDeleteHook != nil {
+		return s.PostDeleteHook(s.Scope, resultErr)
+	}
+	return resultErr
+}
+
+// Pause implements azure.Pauser.
+func (s *Service[T, S]) Pause(ctx context.Context) error {
+	var _ azure.Pauser = (*Service[T, S])(nil)
```

Comment:
I'm curious, why are we doing this type assertion thing here?

Reply:
Normally I'd put this outside the method, but the T and S type parameters aren't defined in that scope. I suppose we could use concrete types and move this outside the method. I can see where this has the marginal benefit of verifying the interface implementation for every set of types a Service is instantiated with, but I doubt that would catch anything new in this case.

Either way, I'm trusting that the compiler optimizes this out completely.

Reply:
If I'd seen this at the module level I wouldn't have blinked, but makes sense why the scope only exists here.